packages/report-server/src/reportBuilder/transform/functions/aggregate/aggregate.ts

```diff
@@ -17,31 +17,73 @@ type AggregateParams = {
   where: (parser: TransformParser) => boolean;
 };
 
-const merge = (mergedRow: Row, newRow: Row, params: AggregateParams): Row => {
-  Object.keys(newRow).forEach((field: string) => {
-    const aggregation = params.getFieldAggregation(field);
-    aggregations[aggregation](mergedRow, field, newRow[field]);
-  });
-  return mergedRow;
+type RowFields = {
+  [fieldKey: string]: FieldValue[];
 };
 
-const aggregate = (rows: Row[], params: AggregateParams): Row[] => {
-  const groupedRows: { [groupKey: string]: Row } = {};
-  const otherRows: Row[] = [];
+type GroupRowFields = {
+  [groupKey: string]: RowFields;
+};
 
+type AggregatedFields = {
+  [fieldKey: string]: FieldValue;
+};
+
+type AggregatedRowFields = {
+  [groupKey: string]: AggregatedFields;
+};
+
+const getGroupRowFields = (
+  rows: Row[],
+  params: AggregateParams,
+): { groupRowFields: GroupRowFields; otherRows: Row[] } => {
+  const groupRowFields: GroupRowFields = {};
   const parser = new TransformParser(rows, functions);
+  const otherRows: Row[] = [];
+
   rows.forEach((row: Row) => {
     if (!params.where(parser)) {
       otherRows.push(row);
       parser.next();
       return;
     }
     const groupKey = params.createGroupKey(row);
-    groupedRows[groupKey] = merge(groupedRows[groupKey] || {}, row, params);
+    groupRowFields[groupKey] = merge(groupRowFields[groupKey] || {}, row);
     parser.next();
   });
 
-  return Object.values(groupedRows).concat(otherRows);
+  return { groupRowFields, otherRows };
+};
+
+const merge = (previousRow: RowFields, newRow: Row): RowFields => {
+  const mergedRow: RowFields = previousRow;
+  Object.keys(newRow).forEach((field: string) => {
+    mergedRow[field] = [...(mergedRow[field] || []), newRow[field]];
+  });
+  return mergedRow;
+};
+
+const getAggregatedRows = (groupRowFields: GroupRowFields, params: AggregateParams): Row[] => {
+  const aggregatedRowFields: AggregatedRowFields = {};
+  Object.keys(groupRowFields).forEach((groupKey: string) => {
+    Object.keys(groupRowFields[groupKey]).forEach((fieldKey: string) => {
+      const aggregation = params.getFieldAggregation(fieldKey);
+      const fieldValue: FieldValue = aggregations[aggregation](groupRowFields[groupKey][fieldKey]);
```

Comment:
```suggestion
      const aggregatedValue = aggregations[aggregation](groupRowFields[groupKey][fieldKey]);
```
Might be worth confirming this, but a lot of the time `Typescript` can infer what type a constant will be when you initialize it, so no need to explicitly give it a type 👍 

As for the variable naming, this is a bit more subjective but personally I'm a big fan of descriptive names instead of 'type-like' names

Reply:
Cool stuff, I did it in this PR for a lot of times, I will delete them.